python/lsst/sims/GalSimInterface/galSimInterpreter.py

```diff
@@ -896,6 +923,75 @@ class carrying all of the information for the object whose image
         self.write_checkpoint()
         return outputString
 
+    def maybeSwitchPSF(gsObject, obj, fft_sb_thresh, pixel_scale=0.2):
+        """
+        Check if the maximum surface brightness of the object is high enough that we should
+        switch to using an fft method rather than photon shooting.
+
+        When we do this, we also switch the PSF model to something slightly simpler with
+        roughly the same wings, but not as complicated in the central core.  Thus, this
+        should only be done when the core is going to be saturated anyway, so we only really
+        care about the wings of the PSF.
+
+        Parameters
+        ----------
+        gsObject: GalSimCelestialObject
+            This contains the information needed to construct a
+            galsim.GSObject convolved with the desired PSF.
+        obj: galsim.GSObject
+            The current GSObject to draw, which might need to be modified
+            if we decide to switch to fft drawing.
+        fft_sb_thresh: float
+            The surface brightness (photons/pixel) where we will switch from
+            photon shooting to drawing with fft if any pixel is above this.
+            Should be at least the saturation level, if not higher.
+        pixel_scale: float [0.2]
+            The CCD pixel scale in arcsec.
+
+        Returns
+        -------
+        galsim.GSObj, bool: obj = the object to actually use
+                            use_fft = whether to use fft drawing
+        """
+        if not fft_sb_thresh:
+            return obj, False
+
+        # obj.original should be a Convolution with the PSF at the end.  Extract it.
+        geom_psf = obj.original.obj_list[-1]
+        all_but_psf = obj.original.obj_list[:-1]
+        try:
+            # If geom_psf is a PhaseScreenPSF, then make a simpler one the just convolves
+            # a Kolmogorov profile with an OpticalPSF.
+            opt_screen = [s for s in geom_psf.screen_list if isinstance(s, galsim.OpticalScreen)][0]
+            optical_psf = galsim.OpticalPSF(
+                    lam=geom_psf.lam,
+                    diam=opt_screen.diam,
+                    aberrations=opt_screen.aberrations,
+                    annular_zernike=opt_screen.annular_zernike,
+                    obscuration=opt_screen.obscuration,
+                    gsparams=geom_psf.gsparams)
+            r0_500 = geom_psf.r0_500_effective
+            atm_psf = galsim.Kolmogorov(lam=geom_psf.lam, r0_500=r0_500,
+                                        gsparams=geom_psf.gsparams)
+            fft_psf = [optical_psf, atm_psf]
+        except AttributeError:
```

Comment:
What are you trying to protect against here?  Maybe add IndexError in case screens don't include an OpticalScreen?

Reply:
I just ran this and did get an `IndexError`:
```
Traceback (most recent call last):
  File "/home/imSim/bin/imsim.py", line 91, in <module>
    image_simulator.run(processes=args.processes)
  File "/home/imSim/python/desc/imsim/ImageSimulator.py", line 250, in run
    results.append(simulate_sensor(self.gs_obj_dict[det_name]))
  File "/home/imSim/python/desc/imsim/ImageSimulator.py", line 370, in __call__
    fft_sb_thresh=fft_sb_thresh)
  File "/home/sims_GalSimInterface/python/lsst/sims/GalSimInterface/galSimInterpreter.py", line 838, in drawObject
    obj, use_fft = self.maybeSwitchPSF(gsObject, obj, fft_sb_thresh)
  File "/home/sims_GalSimInterface/python/lsst/sims/GalSimInterface/galSimInterpreter.py", line 966, in maybeSwitchPSF
    opt_screen = [s for s in geom_psf.screen_list if isinstance(s, galsim.OpticalScreen)][0]
IndexError: list index out of range
```

Reply:
Actually, I forgot about the option of there not being an OptWF.  I was trying to guard against the psf not being a PhaseScreenPSF.  I restructured the code here a bit to (a) only catch an error on the `geom_psf.screen_list` and (b) explicitly handle the case of no OptWF.  b86aa22

Reply:
BTW, @jchiang87 I suspect this error is what caused the problem you reported [here](https://github.com/LSSTDESC/imSim/pull/182#issuecomment-441125329).  Let me know if this fixes it or if it is still happening for you.

Reply:
Using these updates (along with master of GalSim), I now get this error:
```
Traceback (most recent call last):
  File "/global/cscratch1/sd/jchiang8/desc/dev/imSim/bin/imsim.py", line 91, in <module>
    image_simulator.run(processes=args.processes)
  File "/global/cscratch1/sd/jchiang8/desc/dev/imSim/python/desc/imsim/ImageSimulator.py", line 250, in run
    results.append(simulate_sensor(self.gs_obj_dict[det_name]))
  File "/global/cscratch1/sd/jchiang8/desc/dev/imSim/python/desc/imsim/ImageSimulator.py", line 370, in __call__
    fft_sb_thresh=fft_sb_thresh)
  File "/global/cscratch1/sd/jchiang8/desc/dev/sims_GalSimInterface/python/lsst/sims/GalSimInterface/galSimInterpreter.py", line 838, in drawObject
    obj, use_fft = self.maybeSwitchPSF(gsObject, obj, fft_sb_thresh)
  File "/global/cscratch1/sd/jchiang8/desc/dev/sims_GalSimInterface/python/lsst/sims/GalSimInterface/galSimInterpreter.py", line 989, in maybeSwitchPSF
    r0_500 = geom_psf.r0_500_effective
AttributeError: 'PhaseScreenPSF' object has no attribute 'r0_500_effective'
```